File: day7/part2/main.py
# Read input
with open("input.txt") as f:
    grid = [line.rstrip("\n") for line in f]

# Find row containing 'S'
curr_row = next(r for r, row in enumerate(grid) if 'S' in row)

# Beams in current row: {column_index: count}
beams_current_row = {grid[curr_row].index('S'): 1}

splits = 0
timelines = 1
R = len(grid)

while curr_row < R:
    beams_next_row = {}

    def add(col, count):
        beams_next_row[col] = beams_next_row.get(col, 0) + count

    for col, count in beams_current_row.items():
        if grid[curr_row][col] == '.':
            add(col, count)
        else:  # splitter '^'
            add(col - 1, count)
            add(col + 1, count)
            splits += 1
            timelines += count

    beams_current_row = beams_next_row
    curr_row += 1

print("Total timelines:", timelines)
print("Total splits:", splits)

File: day7/part2/main.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    // Read input
    ifstream fin("input.txt");
    vector<string> grid;
    string line;
    while (getline(fin, line)) grid.push_back(line);
    fin.close();

    int R = grid.size();
    int curr_row = 0;

    // Find the row containing 'S'
    for (int r = 0; r < R; r++) {
        if (grid[r].find('S') != string::npos) {
            curr_row = r;
            break;
        }
    }

    // Map of beams in current row: column -> count
    map<int, long long> beamsCurrentRowMap;
    beamsCurrentRowMap[grid[curr_row].find('S')] = 1;

    long long splits = 0;
    long long timelines = 1;

    while (curr_row < R) {
        map<int, long long> beamsNextRowMap;

        auto add = [&](int col, long long count) {
            beamsNextRowMap[col] += count;
        };

        for (auto &[col, count] : beamsCurrentRowMap) {
            if (grid[curr_row][col] == '.') {
                add(col, count); // continues straight
            } else { // splitter '^'
                add(col - 1, count);
                add(col + 1, count);
                splits++;
                timelines += count;
            }
        }

        beamsCurrentRowMap.swap(beamsNextRowMap);
        curr_row++;
    }

    cout << "Total timelines: " << timelines << endl;
    cout << "Total splits: " << splits << endl;

    return 0;
}
